explainer.md

```diff
@@ -139,20 +176,66 @@ let depthValueInMeters = depthInfo.getDepth(x, y);
 
 ## Appendix: Proposed Web IDL
 
+The comments are left in place to provide additional context on the API to the readers of the IDL and should also be repeated in the explainer text above.
+
 ```webidl
+enum XRDepthSensingUsage {
+  "cpu-optimized",
+  "gpu-optimized"
+};
+
+enum XRDepthFormat {
+  // Has to be supported everywhere:
+  "luminance-alpha", // internal_format = LUMINANCE_ALPHA, type = UNSIGNED_BYTE,
+                     // 2 bytes per pixel, unpack via dot([R, A], [255.0, 256*255.0])
+  // Other formats do not have to be supported by the user agents:
+  "float32",         // internal_format = R32F, type = FLOAT
+};
+
+// Post-session-creation, the depth sensing state needs to be configured
+// with the desired data format.
+dictionary XRDepthSensingStateInit {
+  XRDepthSensingUsage usage = "cpu-optimized";
+  XRDepthFormat depthFormat = "luminance-alpha";
+};
+
+partial interface XRSession {
+  // Non-null iff depth-sensing is enabled:
+  readonly attribute XRDepthFormat? preferredDepthFormat;
+  readonly attribute XRDepthSensingUsage? preferredDepthSensingUsage;
+
+  // Can be called iff depth-sensing is enabled, will throw if unsupported format is passed in.
+  // The user agent can reject session creation if the desired usage is not supported.
+  // The user agent can support reconfiguring a session, but that is not required.
+  Promise<void> updateDepthSensingState(optional XRDepthSensingStateInit state = {});
+};
+
+partial interface XRFrame {
+  // Throws if depth sensing state was never updated,
+  // returns null if the view does not have a corresponding depth buffer.
+  XRDepthInformation? getDepthInformation(XRView view);
+};
+
+[SecureContext, Exposed=Window]
 interface XRDepthInformation {
-  readonly attribute Uint16Array data;
+  // All methods / attributes are accessible only when the frame
+  // that the depth information originated from is active and animated.
 
+  [SameObject] readonly attribute DataView data;
   readonly attribute unsigned long width;
   readonly attribute unsigned long height;
 
-  [SameObject] readonly attribute XRRigidTransform normTextureFromNormView;
-
   float getDepth(unsigned long column, unsigned long row);
+
+  // Must be valid irrespective of initialization hint:
+  [SameObject] readonly attribute XRRigidTransform normTextureFromNormView;
+  readonly attribute float rawValueToMeters;
 };
 
-partial interface XRFrame {
-  XRDepthInformation getDepthInformation(XRView view);
+partial interface XRWebGLBinding {
+  // Must succeed when the depth API was initialized as "gpu-optimized". Otherwise, must throw.
+  // Returns opaque texture, its format is determined by the current depth sensing state.
+  WebGLTexture? getDepthTexture(XRDepthInformation depthInformation);
```

Comment:
Why pass depthInformation? Wouldn't XRView make more sense?

Reply:
Hm, `XRView` should also work. I think I prefer to keep `XRDepthInformation` here since the texture is not useful w/o it anyway, and this way it's clear that `XRWebGLBinding` "extends" the depth information in a way. Also, from spec editing perspective, we already need to tie the usefulness of `XRDepthInformation` to an `XRFrame` from which it originated (the `XRDepthInformation` is only usable as long as the frame it's from is `animated` and `active`), so we'd have to impose similar restrictions, & `XRView` currently does not know about the `XRFrame` it's coming from. But those are all spec editing issues that can be solved and we should just go with the API that makes more sense here.

Reply:
We need an XRView unless the texture is side-by-side

Reply:
Not sure if I understand? We'll need to have some way of describing that the depth data is available on some object in the spec, it should not be difficult to associate this data with either `XRDepthInformation` or `XRView` and spec the algorithm so that it says sth. like "the returned texture must contain the depth data associated with passed in XRView/XRDepthInformation".

One more thing: if we make it available on `XRView`, it could send an incorrect message to the developers that any `XRView` can be used to get depth data - this is not the case, the depth API needs to be first enabled & configured. If you have `XRDepthInformation`, it means you already went through those steps. If I could, I would put a `WebGLTexture texture` attribute attribute on the `XRDepthInformation`, but since I cannot, this seemed to me like the second best thing to do. :)

Reply:
I see. I had missed that `XRDepthInformation` was returned by `getDepthInformation`

> If I could, I would put a WebGLTexture texture attribute attribute on the XRDepthInformation, but since I cannot, this seemed to me like the second best thing to do. :)

Maybe you can move the depth texture into `XRDepthInformation` and `getDepthInformation` to `XRWebGLBinding`?
If the depth texture can be side-by-side, you would also need to return a viewport that describes what portion of the texture should be used.

Reply:
Moving `getDepthInformation()` into `XRWebGLBinding` would tie the entire depth API to WebGL, which is not something we should do here I think - for one, we still have a CPU path that does not need WebGL, and we also know WebGPU is upcoming, so the API should be able to integrate with it as well (likely via sth like a `XRWebGPUBinding` or something similar). As for what portion of the texture should be used, I imagined the entire texture will always encompass the entire view's viewport, so we should not have to provide additional viewport.

Reply:
> Will there be experiences that use depth but no WebGL?

So far I have not seen any requests to make it possible, but since we want to have a mode of operation that does not need to require the use of WebGL, I think we should not tie the entire API surface to WebGL and/or WebGPU.

> Maybe we could make it so `getDepthInformation` on XRFrame returns an object for the CPU option and `getDepthInformation` on `XRWebGLBinding` return an object for the GPU option?

We'd actually need to have something like:
XRDepthInformation = generic info for all types (size, UV transform, raw value to meters scale factor)
XRCPUDepthInformation public XRDepthInformation = add info for CPU (`ArrayBufferView data` attribute, `float getDepth(x,y)` method), obtainable from `XRFrame.getDepthInformation(XRView view)`
XRWebGLDepthInformation public XRDepthInformation = add info for WebGL (`WebGLTexture texture` member that is an opaque texture defined by layers spec draft), obtainable from `XRWebGLBinding.getDepthInformation(XRView view)`
XRWebGPUDepthInformation public XRDepthInformation = add info for WebGPU (`GPUTexture texture` member?), obtainable from `XRWebGPUBinding.getDepthInformation(XRView view)`?

For context, I wanted the `XRDepthInformation` to serve as a handle that could be passed to the WebGL/WebGPU bindings to retrieve the corresponding GPU object (WebGLTexture/GPUTexture). This way, the path that the developers need to take is potentially clearer:
"XRWebGLBinding takes XRDepthInformation, that is obtainable from XRFrame when I pass XRView, so I need to be in a rAFcb, enumerate the views I'm interested in, grab the XRDepthInformation (that I also need for the UV transform) and ask the binding for a texture. Oh, and I can obtain XRDepthInformation only if the session has 'depth-sensing' enabled."

The main problem that I see with the handle approach is that it _sometimes_ would also contain the CPU-accessible data (the `data` member), so just because of that it may make more sense to have separate types to represent the different possible modes of accessing the data.

Reply:
Resolving, see latest version.